x-pack/plugin/esql/compute/build.gradle

```diff
@@ -365,6 +365,27 @@ tasks.named('stringTemplates').configure {
     it.inputFile =  arrayStateInputFile
     it.outputFile = "org/elasticsearch/compute/aggregation/DoubleArrayState.java"
   }
+  File valuesAggregatorInputFile = new File("${projectDir}/src/main/java/org/elasticsearch/compute/aggregation/X-ValuesAggregator.java.st")
+  template {
+    it.properties = intProperties
+    it.inputFile =  valuesAggregatorInputFile
+    it.outputFile = "org/elasticsearch/compute/aggregation/ValuesIntAggregator.java"
+  }
+  template {
+    it.properties = longProperties
+    it.inputFile =  valuesAggregatorInputFile
+    it.outputFile = "org/elasticsearch/compute/aggregation/ValuesLongAggregator.java"
+  }
+  template {
+    it.properties = doubleProperties
+    it.inputFile =  valuesAggregatorInputFile
+    it.outputFile = "org/elasticsearch/compute/aggregation/ValuesDoubleAggregator.java"
+  }
+  template {
+    it.properties = bytesRefProperties
+    it.inputFile =  valuesAggregatorInputFile
+    it.outputFile = "org/elasticsearch/compute/aggregation/ValuesBytesRefAggregator.java"
+  }
```

Comment:
I'm using string template to build the per-type aggregator behavior because it's fairly copy-and-paste-y. And the `X-ValuesAggregator.java.st` file does a good job showing how the types differ and are the same. Those generated files are then processed by the annotation processing code to make the actual aggregators. So code generation to power code generation.